ncm-opennebula/src/main/perl/OpenNebula/Server.pm

```diff
@@ -0,0 +1,317 @@
+
+Detects C<OpenNebula> version through opennebula-server probe files,
+the value gathered from the file must be untaint.
+
+=cut
+
+sub detect_opennebula_version
+{
+    my ($self) = @_;
+
+    my $fh = CAF::FileReader->new($OPENNEBULA_VERSION_FILE, log => $self);
+    if (! "$fh") {
+        $self->error("Not found OpenNebula version file: $OPENNEBULA_VERSION_FILE");
+        return;
+    };
+
+    my $version;
+    my $msg = '';
+    # untaint value
+    if ("$fh" =~ m/^(\d+\.\d+(?:\.\d+)?$)/m ) {
+        local $@;
+        eval {
+            $version = version->new($1);
+        };
+        $msg = "$@";
+    } else {
+        $msg = "No match for version regexp";
+    }
+
+    if ($version) {
+        $self->verbose("OpenNebula $OPENNEBULA_VERSION_FILE file has version $version.");
+        return $version;
+    } else {
+        $self->error("No valid version available from $OPENNEBULA_VERSION_FILE file. $msg");
+        return;
+    };
+}
+
+=item change_opennebula_passwd
+
+Sets a new C<OpenNebula> service password.
+
+=cut
+
+sub change_opennebula_passwd
+{
+    my ($self, $user, $passwd) = @_;
+
+    my $cmd;
+    if ($user eq $SERVERADMIN_USER) {
+        $cmd = [$user, join(' ', '--driver server_cipher', $passwd)];
+    } else {
+        $cmd = [$user, $passwd];
+    };
+    my $output = $self->run_oneuser_as_oneadmin_with_ssh($cmd, "localhost", 1);
+    if (!$output) {
+        $self->error("Quattor unable to modify current $user passwd.");
+        return;
+    } else {
+        $self->info("$user passwd was set correctly.");
+    }
+    $self->set_one_auth_file($user, $passwd);
+    return 1;
+}
+
+=item set_one_service_conf
+
+Sets C<OpenNebula> configuration files used by
+the deamons, if the configuration file is changed the
+service must be restarted afterwards.
+
+=cut
+
+sub set_one_service_conf
+{
+    my ($self, $data, $service, $config_file, $cfggrp) = @_;
+
+    my $cfgv = $self->detect_opennebula_version;
+    if ($cfgv >= version->new("5.0.0") and $service eq 'oned') {
+        $self->verbose("Found OpenNebula >= 5.0 configuration flag");
+        $data->{datastore_mad}->{arguments} = $ONED_DATASTORE_MAD;
+    };
+    my $oned_templ = $self->process_template($data, $service);
+    my %opts = $self->set_file_opts();
+    return if ! %opts;
+    $opts{group} = $cfggrp if ($cfggrp);
+    my $fh = $oned_templ->filewriter($config_file, %opts);
+    my $status = $self->is_conf_file_modified($fh, $config_file, $service, $oned_templ);
+
+    return $status;
+}
+
+=item is_conf_file_modified
+
+Checks C<OpenNebula> configuration file status.
+
+=cut
+
+sub is_conf_file_modified
+{
+    my ($self, $fh, $file, $service, $data) = @_;
+
+    if (!defined($fh)) {
+        if (defined($service) && defined($data)) {
+            $self->error("Failed to render $service file: $file (".$data->{fail}."). Skipping");
+        } else {
+            $self->error("Problem rendering $file");
+        }
+        $fh->cancel();
```

Comment:
cannot work if `$fh` is not defined. but be careful, the `cancel` is important, it prevents that changes are made